Markdown mention bug

m-e-l-u-h-a-n opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
Fixes: #17535

**Testing plan:** <!-- How have you tested? -->
Testing is done in development server and by adding new markdown test for testing changes.

**GIFs or screenshots:**
**For user mention:**
```
source:  @**Cordelia Lear**  @**invalid** and @**King Hamlet**
```
output:
![image](https://user-images.githubusercontent.com/63504956/110690247-b7820480-8209-11eb-9c69-853f6e363023.png)

**For user group mention:**
```
source:  @*hamletcharacters*  @*invalid group* @*hamletcharacters*
```
output:
![image](https://user-images.githubusercontent.com/63504956/110690738-4b53d080-820a-11eb-82e2-0a4f09633dae.png)

**For stream mention:**
```
source:  #**invalid stream>first BOTS wasn't loading**, #**devel>invalid topic**, #**invalid stream>invalid topic**  #**devel>first BOTS wasn't loading**  and then #**support**
```
output:
![image](https://user-images.githubusercontent.com/63504956/110690983-9372f300-820a-11eb-8793-62e1ed4fbfbd.png)

timabbott commented on zerver/lib/markdown/__init__.py:
```diff
@@ -1829,13 +1829,19 @@ def handleMatch(self, m: Match[str], data: str) -> Tuple[Optional[Element], Any,
+            # coming after (?<![^\s\'\"\(,:<]), so @ is also
+            # included in match so it needs to be strip.
+            if match.startswith("@*") and match.endswith("*"):
+                name = match[2:-1]
```
I think it'd be better to write this with variables, with `start_syntax = "@*"`, and then using that variable (E.g. len(start_syntax):-len(end_syntax)`) here -- we try to avoid numbers like this.

m-e-l-u-h-a-n replied:
Made these changes!

timabbott commented on zerver/lib/markdown/__init__.py:
```diff
@@ -125,27 +125,22 @@ def normal_compile(pattern: str) -> Any:
+# ends by double asterisks
+STREAM_TOPIC_LINK_REGEX = (
+    r"(?<![^\s'\"\(,:<])\#\*\*(?P<stream_name>[^\*>]+)>(?P<topic_name>[^\*]+)\*\*"
+)
```
What's the thinking around removing the line-wrapping on these?

m-e-l-u-h-a-n replied:
oops, sorry I might have changed this during several approaches I tried(that required me to compile without using `VERBOSE` flag) and later forgot to change it back. I have fixed it.

timabbott commented on zerver/lib/markdown/__init__.py:
```diff
@@ -125,27 +125,22 @@ def normal_compile(pattern: str) -> Any:
-
-@one_time
-def get_compiled_stream_link_regex() -> Pattern:
-    return verbose_compile(STREAM_LINK_REGEX)
```
@showell do you remember the context around adding this function in ad071ced4782aba135b59db8597fa4f75f0e91a2?  I think it was part of a performance optimization effort, and we should verify that we're not regressing performance by removing that pre-compiling.

m-e-l-u-h-a-n replied:
I have updated this so that we are still pre-compiling the regex to avoid concerned regressions related to this. But `verbose_compile` is no longer used for pre-compiling `STREAM_LINK_REGEX`, `STREAM_TOPIC_LINK_REGEX` as it adds ^(.*?) and (.*?)$ which cause extra overhead of matching pattern which is not required with new `InlineProcessor `[(docs mentioning this)](https://python-markdown.github.io/extensions/api/#comparison-with-new-inlineprocessor) these extra patterns at beggining and end are not required. So, `StreamPattern` and `StreamTopicPattern` now define their own `__init__` method for accepting pre-compiled the regex.
As we cannot use `CompiledPattern` class for this(because it inherits `markdown.inlinepatterns.Pattern` which is older API and reason for this bug), so I had to define their separate `__init__` constructor.

m-e-l-u-h-a-n commented:
@timabbott @andersk I have updated the pr and addressed all the comments from Tim explaining changes done for them comment wise. Also I have cleaned up code and added some comments explaining the changes where ever necessary. It now passes CI and is ready for review.

m-e-l-u-h-a-n commented:
I think if we migrate `CompiledPattern` to inherit from `markdown.inlinepatterns.InlineProcessor` instead of markdown.inlinepatterns.Pattern then, their we would be no need to define separate constructors(doing same work of accepting pre-compiled regex) for each of `StreamPattern` and `StreamTopicPattern`. Would this be a good idea?

timabbott commented on zerver/lib/markdown/__init__.py:
```diff
@@ -1833,13 +1833,25 @@ def handleMatch(  # type: ignore[override] # supertype incompatible with supersu
+            if match.startswith(start_syntax) and match.endswith(end_syntax):
+                name = match[len(start_syntax) : -len(end_syntax)]
+            else:
+                name = match
```
Is this `else` clause possible?  I don't understand under what circumstance it would be, given the regular expression.

(If it is, do we have a test for it?).

timabbott reviewed (commented):
Thanks for working on this @m-e-l-u-h-a-n!  It looks pretty good; I posted one comment, and also pushed back after tweaking the last commit to auto-close the issue when we merge this.

m-e-l-u-h-a-n commented:
@timabbott I made changes in `UserGroupMentionPattern` to not reqiure else block also it was not required with current regex. Also I posted on [CZO](https://chat.zulip.org/#narrow/stream/9-issues/topic/markdown.20mention.20bug/near/1132863) about your question on migration of `CompiledPattern`. Can you please give it a look.

timabbott commented on zerver/lib/markdown/__init__.py:
```diff
@@ -1833,13 +1833,23 @@ def handleMatch(  # type: ignore[override] # supertype incompatible with supersu
+            # As InlineProcessor API sends all the matching part
+            # coming after (?<![^\s\'\"\(,:<]), so @ is also
+            # included in name so it needs to be strip.
+            if name.startswith(start_syntax) and name.endswith(end_syntax):
```
Should this be an `assert` rather than an `if`?

m-e-l-u-h-a-n replied:
The else for this is never expected to happen with current regex and `InlineProcessor` api. So if we do not put assert here this it will be easily spotted as a bug if any thing else ever happens.

timabbott commented on zerver/lib/markdown/__init__.py:
```diff
@@ -1833,13 +1833,23 @@ def handleMatch(  # type: ignore[override] # supertype incompatible with supersu
-            name = extract_user_group(match)
+            # As InlineProcessor API sends all the matching part
+            # coming after (?<![^\s\'\"\(,:<]), so @ is also
+            # included in name so it needs to be strip.
```
"needs to be stripped" would be normal English.

m-e-l-u-h-a-n replied:
made these changes!

andersk commented on zerver/lib/markdown/__init__.py:
```diff
@@ -1826,17 +1845,27 @@ def handleMatch(self, m: Match[str]) -> Optional[Element]:
                 el.set("class", "user-mention")
                 text = f"@{text}"
             el.text = markdown.util.AtomicString(text)
-            return el
-        return None
-
+            return el, match_span[0], match_span[1]
+        return None, None, None
 
-class UserGroupMentionPattern(markdown.inlinepatterns.Pattern):
-    def handleMatch(self, m: Match[str]) -> Optional[Element]:
-        match = m.group(2)
 
+class UserGroupMentionPattern(markdown.inlinepatterns.InlineProcessor):
+    def handleMatch(  # type: ignore[override] # supertype incompatible with supersupertype
+        self, m: Match[str], data: str
+    ) -> Union[Tuple[None, None, None], Tuple[Element, int, int]]:
+        name = m.group(0)
+        match_span = m.span()
         db_data = self.md.zulip_db_data
+
+        # As patterns like @*group name* are captured.
+        start_syntax = "@*"
+        end_syntax = "*"
         if self.md.zulip_message and db_data is not None:
-            name = extract_user_group(match)
+            # As InlineProcessor API sends all the matching part
+            # coming after (?<![^\s\'\"\(,:<]), so @ is also
+            # included in name so it needs to be strip.
+            if name.startswith(start_syntax) and name.endswith(end_syntax):
+                name = name[len(start_syntax) : -len(end_syntax)]
```
Why not `name = extract_user_group(m[1])`?

m-e-l-u-h-a-n replied:
extract_user_group(m[1]) will perform name = name[1:-1], but we have name="@*caught_pattern*" initially, so we need name[2:-1]. So I extracted it here instead of making a new function, or putting a flag in the current function.

andersk replied:
Note `m[1]`, not `m[0]`. Group 1 is the part *after* the `@`. (That’s why `extract_user_group` works at all.)

m-e-l-u-h-a-n replied:
Made these changes.

andersk commented on zerver/lib/markdown/__init__.py:
```diff
@@ -1845,32 +1874,42 @@ def handleMatch(self, m: Match[str]) -> Optional[Element]:
-class StreamPattern(CompiledPattern):
+class StreamPattern(markdown.inlinepatterns.InlineProcessor):
+    def __init__(self, compiled_re: Pattern, md: markdown.Markdown) -> None:
+        # This is similar to the superclass's small __init__ function,
+        # but we skip the compilation step and let the caller give us
+        # a compiled regex.
+        self.compiled_re = compiled_re
```
That shouldn’t be necessary. `re.compile` already accepts a compiled regex and skips the compilation step.

m-e-l-u-h-a-n replied:
superclass constructor calls `re.compile` with `DOTALL` and `UNICODE` flag which fail if the pattern is already compiled so I had to define the constructor in order to accept pre-compiled pattern.

andersk commented on zerver/lib/markdown/__init__.py:
```diff
@@ -1826,17 +1845,27 @@ def handleMatch(self, m: Match[str]) -> Optional[Element]:
-            return el
-        return None
-
+            return el, match_span[0], match_span[1]
```
`return el, m.start(), m.end()` seems a little cleaner and avoids the separate `span()` call.

m-e-l-u-h-a-n replied:
made these changes.

andersk commented on zerver/lib/markdown/__init__.py:
```diff
@@ -135,7 +135,15 @@ def normal_compile(pattern: str) -> Any:
+    # With new InlineProcessor these extra patterns
+    # are not required.
+    return re.compile(
+        f"{STREAM_LINK_REGEX}",
```
`f"{STREAM_LINK_REGEX}"` is just `STREAM_LINK_REGEX`.

m-e-l-u-h-a-n replied:
Yeah I made these changes.

m-e-l-u-h-a-n commented:
I have made the changes as suggested. As changes were minor so I have replied in respective comments about the changes done. This pr is ready for another review. Please have a look.

m-e-l-u-h-a-n commented:
@andersk I have made the changes as suggested. Can you please give it a review.

timabbott commented:
This is great, merged, thanks @m-e-l-u-h-a-n!  I hope the series of iterations on this gives you a feel for our approach of thinking through the details to try to get to a nice implementation that's readable and correct.
